File: PE#10/PE#10sieve.py
"""
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
Find the sum of all the primes below two million.
"""

def prime_sum():
  prime = [True]*2_000_000
  for i in range(2, 2_000_000):
    if prime[i]:
      for j in range(i*i, 2_000_000, i):
        prime[j]=False
  sum=0
  for i in range(2, 2_000_000):
    if prime[i]:sum=sum+i
  print(sum)

prime_sum()

File: PE#10/PE#10sieve.cpp
#include <iostream>
#include <vector>
using namespace std;

// Reference : https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes

void PrimeSum() {
  vector<bool> prime(2e6, true);
  for (size_t i = 2; i < 2e6; ++i) {
    if (prime[i]) {
      for (size_t j = i * i; j < 2e6; j = j + i) prime[j] = false;
    }
  }

  uint64_t sum = 0;
  for (size_t i = 2; i < 2e6; ++i) {
    if (prime[i]) sum = sum + i;
  }
  cout << sum << endl;
}

int main() { PrimeSum(); }
